### PR TITLE
Fix the issues in `DefaultCertValidator::doVerifyCertChain` and `ContextImpl::verifyCertChain`

### PR DESCRIPTION
**Pull request by tyxia**
1. In DefaultCertValidator::doVerifyCertChain() function, there is one corner case not handled right now : `trusted_ca = false` and `no hash or SAN matcher` . Basically, it means no validation context is provided so no certificate validation is performed. Current code will still return `OK` for this case. `NotValidated` seems to be an initial state that needs to be updated to either `Failed` or `Validated` eventually. If it stay as `NotValidated`, it should be treated as `Error`.
2. It reveals one bug via sds_dynamic_integration_test's QUIC path(The only failure caused by change above. TCP path of this test works fine).
Basically, the problem is QUIC client side register  and then will invoke the  `verifyCertificate()` callback because its cert_verify_mode is set to `SSL_VERIFY_PEER` but no validation_context is provided. Thus, it will run into `NotValidated` case I described in 1 and my change return it as `ERROR`.

Similar to [openSSL logic](https://github.com/google/boringssl/blob/e9c5d72c09e01a0f71f30f7c3454e5e7f8711476/ssl/ssl_x509.cc#L409), the error can be treated as non-fatal if mode is SSL_VERIFY_NONE. The proposed fix is in ContextImpl::verifyCertChain( ) (which is only used by QUIC path): Query the true verify_mode that is initialized in `DefaultCertValidator::initializeSslContexts`(it will be `SSL_VERIFY_NONE` for this case because there is no validation_context config provided). 
(PS: I discussed with QUIC folk, QUIC test could be investigated as well. But I feel the proposed fix is needed regardless)

Signed-off-by: Tianyu Xia tyxia@google.com

Risk Level: Low
Testing: Local test

### CODE REVIEW

**Comment by alyssawilk** [comment]
Just to be aware, we've been making some improvements to Envoy tooling for more prompt reviews, but draft PRs are not included in the tooling.  I'm updating the contribution guidelines to make that clear (https://github.com/envoyproxy/envoy/pull/17063/files) , but as this PR predates the repokitteh warning I just want to call out that draft PRs are likely not going to get fast turnaround. Please consider marking this as ready for review if it stalls, and you want the assignee to take a look!

**Comment by tyxia** [comment]
> Just to be aware, we've been making some improvements to Envoy tooling for more prompt reviews, but draft PRs are not included in the tooling. I'm updating the contribution guidelines to make that clear (https://github.com/envoyproxy/envoy/pull/17063/files) , but as this PR predates the repokitteh warning I just want to call out that draft PRs are likely not going to get fast turnaround. Please consider marking this as ready for review if it stalls, and you want the assignee to take a look!

Thank you very much for information and pointers! I created this draft PR intentionally to trigger the CI run (as the sanity check):). I will mark it as ready-for-review when when it is ready. Please let me know if you have any other instructions. Thanks!

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 1]
```diff
@@ -216,7 +216,8 @@ int DefaultCertValidator::doVerifyCertChain(
   }
 
   return allow_untrusted_certificate_ ? 1
-                                      : (validated != Envoy::Ssl::ClientValidationStatus::Failed);
+         : verify_trusted_ca_         ? validated != Envoy::Ssl::ClientValidationStatus::Failed
```
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:219]: style: please don't nest `? :`, it's hard to read

Comment (tyxia) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc]: Thanks for the review! Acked and updated.

**Review by rojkov — APPROVED** [review]
Thanks! LGTM

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 2]
```diff
@@ -215,8 +215,15 @@ int DefaultCertValidator::doVerifyCertChain(
     }
   }
 
-  return allow_untrusted_certificate_ ? 1
-                                      : (validated != Envoy::Ssl::ClientValidationStatus::Failed);
+  // If `trusted_ca` exists, it is already verified in the code above. Thus, we just need to make
+  // sure the verification for other validation context configurations doesn't fail (i.e. either
+  // `NotValidated` or `Validated`). If `trusted_ca` doesn't exist, we will need to make sure other
+  // configurations are verified and the verification succeed.
+  int validation_status = verify_trusted_ca_
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:222]: `bool`?

Comment (tyxia) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:222]: Thanks for review! I think either bool or int will work here as bool <-> int implicit conversation will be triggered.
The reason I used int here is to align with value 1 in ternary operator in line 226 and the return type of this function (which is int).

Please let me know what do you think.

Comment (yanavlasov) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:222]: I think `int` is fine given that the function returns the same error code as `X509_verify_cert`

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**